### PR TITLE
Add resource limiting to the Wasmtime API.

### PR DESCRIPTION
**Pull request by peterhuene**
This PR adds a `ResourceLimiter` trait to the Wasmtime API.

When used in conjunction with `Store::new_with_limiter`, this can be used to
monitor and prevent WebAssembly code from growing linear memories and tables.

This is particularly useful when hosts need to take into account host resource
usage to determine if WebAssembly code can consume more resources.

A simple `Limiter` is also included with these changes that will
simply limit the size of linear memories or tables for all instances created in
the store based on static values.

### CODE REVIEW

**Comment by peterhuene** [comment]
I'm marking this as a draft PR as I'd like feedback on this approach before I create a formal RFC for the API changes (I find it useful to discuss code rather than markdown).

**Comment by peterhuene** [comment]
So I've seen this CI failure and I think it's a GitHub actions things.  As far as I can tell, we're using `ubuntu-latest` for all of the Linux CI, but for some reason we're occasionally getting Ubuntu 20.04 instead of what *should* be the current latest, which is 18.04 (according to [this](https://github.com/actions/virtual-environments)).

Although the GitHub actions virtual environment for Ubuntu 20.04 says Clang 9 is installed, perhaps lldb-9 is not and thus the expected `LLDB` path set for the tests (`/usr/bin/lldb-9`) doesn't work.

**Comment by peterhuene** [comment]
So it appears that 20.04 is now `ubuntu-latest` by design, so we were caught in an upgrade to GitHub actions.

File: crates/runtime/src/instance.rs [hunk 1]
```diff
@@ -378,11 +412,21 @@ impl Instance {
...
+            .unwrap_or_else(|| panic!("no memory for index {}", memory_index.index()));
+
+        if let Some(limiter) = &self.limiter {
+            let current = memory.size();
+            let desired = current.checked_add(delta)?;
+
+            if !limiter.memory_growing(current, desired, memory.maximum()) {
+                return None;
+            }
+        }
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs:427]: Would it be possible to store the `limiter` into the `Memory` itself rather than in the `Instance`?

File: crates/runtime/src/instance/allocator.rs [hunk 2]
```diff
@@ -59,6 +59,9 @@ pub struct InstanceAllocationRequest<'a> {
 
     /// The pointer to the stack map registry to use for the instance.
     pub stack_map_registry: *mut StackMapRegistry,
+
+    /// The resource limiter to use for the instance.
+    pub limiter: Option<Arc<dyn ResourceLimiter>>,
```
Comment (alexcrichton) [inline crates/runtime/src/instance/allocator.rs:64]: If this ends up being stored in individual tables/memories then this could probably turn into `&'a Arc<dyn ...>`

File: crates/runtime/src/memory.rs [hunk 3]
```diff
@@ -33,6 +33,10 @@ pub trait RuntimeLinearMemory {
     /// Returns the number of allocated wasm pages.
     fn size(&self) -> u32;
 
+    /// Returns the maximum number of pages the memory can grow to.
+    /// Returns `None` if the memory is unbounded.
+    fn maximum(&self) -> Option<u32>;
```
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:38]: Could the  docs here clarify that this is about the runtime limit maximum of the memory, not the maximum size as prescribed by the type in wasm?

File: crates/wasmtime/src/limiter.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+/// A [`Store`] can be created with a resource limiter so that hosts can take into account
+/// non-WebAssembly resource usage to determine if a linear memory or table should grow.
+pub trait ResourceLimiter {
+    /// Notifies the resource limiter that an instance's linear memory has been requested to grow.
+    ///
+    /// * `current` is the current size of the linear memory in WebAssembly page units.
+    /// * `desired` is the desired size of the linear memory in WebAssembly page units.
+    /// * `maximum` is either the linear memory's maximum or a maximum from an instance allocator,
+    ///   also in WebAssembly page units. A value of `None` indicates that the linear memory is
+    ///   unbounded.
```
Comment (alexcrichton) [inline crates/wasmtime/src/limiter.rs:15]: To make sure I understand this right, if any `maximum` is in play then this field is set to the minimum of all those maximums? The sources of `maximum` are:

* The actual wasm type itself
* The pooling allocator's configured maximum
* A custom implementation of `RuntimeMemory` has a configured maximum

If my understanding is right, could this be clarified a bit? Basically `None` is only returned if literally nothing is limiting the memory growth, but otherwise `Some` doesn't imply that the wasm type itself had a maximum listed, just that something else somewhere is placing a constraint.

Also, presumably, this function isn't called unless `desired` is already known to be less than `maximum`, right? (or is it called even in those cases?)

File: crates/wasmtime/src/limiter.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+impl StaticResourceLimiter {
+    /// Creates a new [`StaticResourceLimiter`].
+    ///
+    /// The `memory_limit` parameter is the number of WebAssembly pages a linear memory can grow to.
+    /// If `None`, the limiter will not limit linear memory growth.
+    ///
+    /// The `table_limit` parameter is the number of elements a table can grow to.
+    /// If `None`, the limiter will not limit table growth.
+    pub fn new(memory_limit: Option<u32>, table_limit: Option<u32>) -> Self {
```
Comment (alexcrichton) [inline crates/wasmtime/src/limiter.rs:95]: This structure looks ripe for expansion to limiting other resources in the future, so perhaps this could use a builder-api pattern to avoid future API breakage?

File: crates/wasmtime/src/limiter.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+pub(crate) struct ResourceLimiterProxy {
+    store: Weak<StoreInner>,
+    limiter: Box<dyn ResourceLimiter>,
+}
+
+impl ResourceLimiterProxy {
+    pub(crate) fn new(store: &Store, limiter: impl ResourceLimiter + 'static) -> Self {
+        Self {
+            store: store.weak(),
+            limiter: Box::new(limiter),
```
Comment (alexcrichton) [inline crates/wasmtime/src/limiter.rs:56]: To avoid double-allocations and double-indirection could `ResourceLimiterProxy` have a type variable perhaps?

File: crates/wasmtime/src/store.rs [hunk 7]
```diff
@@ -98,6 +98,7 @@ pub(crate) struct StoreInner {
     current_poll_cx: Cell<*mut Context<'static>>,
     out_of_gas_behavior: Cell<OutOfGas>,
     context_values: RefCell<HashMap<TypeId, Box<dyn Any>>>,
+    limiter: RefCell<Option<Arc<dyn wasmtime_runtime::ResourceLimiter>>>,
```
Comment (alexcrichton) [inline crates/wasmtime/src/store.rs:101]: It looks like this value doesn't change over time, so I don't think it needs to be backed by a `RefCell`?

File: crates/wasmtime/src/trampoline.rs [hunk 8]
```diff
@@ -77,6 +77,7 @@ fn create_handle(
                     as *const VMExternRefActivationsTable
                     as *mut _,
                 stack_map_registry: store.stack_map_registry() as *const StackMapRegistry as *mut _,
+                limiter: None,
```
Comment (alexcrichton) [inline crates/wasmtime/src/trampoline.rs:80]: I think this needs to be specified? Otherwise are memories created via `Memory::new` (or tables) limited via the limiter?

**Review by alexcrichton — COMMENTED** [review]
We currently also have a [few methods on `Config`](https://docs.rs/wasmtime/0.25.0/wasmtime/struct.Config.html#method.max_instances) for limiting resources, so I think it'd be good to have everything in one place and move those here as well? Otherwise this looks pretty neat to me and seems like a reasonable feature!

File: crates/wasmtime/src/limiter.rs (hunk 4 above)
Comment (peterhuene) [reply crates/wasmtime/src/limiter.rs:15]: Your understanding is correct and I can expand upon it in the docs for clarity.

The function is called even when desired exceeds the maximum as some implementers will want to know when the maximum was reached even if it was not directly limited by the limiter.

If that feels a little strange, perhaps a different function with the intention that it is only called when the memory couldn't grow due to it exceeding the maximum (from wasm / pooling allocator / custom runtime memory)?

File: crates/wasmtime/src/limiter.rs (hunk 5 above)
Comment (peterhuene) [reply crates/wasmtime/src/limiter.rs:95]: Good idea.  I'll implement a builder.

File: crates/wasmtime/src/trampoline.rs (hunk 8 above)
Comment (peterhuene) [reply crates/wasmtime/src/trampoline.rs:80]: Good point.  I'll fix.

File: crates/wasmtime/src/store.rs (hunk 7 above)
Comment (peterhuene) [reply crates/wasmtime/src/store.rs:101]: It currently does as it needs to sit on `inner` which is behind `Rc` and there's a circular reference between the proxy and the `Store`, which is why there is ugliness in `Store::new_with_limiter` to set this thing.

I'm not happy about it at all, so if there's a better way to do this such that the limiter can pass the store along in the callbacks, I'm definitely open to it.

File: crates/runtime/src/memory.rs (hunk 3 above)
Comment (peterhuene) [reply crates/runtime/src/memory.rs:38]: I will expand these docs.

File: crates/runtime/src/instance.rs (hunk 1 above)
Comment (peterhuene) [reply crates/runtime/src/instance.rs:427]: Can do.

File: crates/runtime/src/instance/allocator.rs (hunk 2 above)
Comment (peterhuene) [reply crates/runtime/src/instance/allocator.rs:64]: I'll change this, no reason to clone it for the request until a memory / table needs it.

File: crates/wasmtime/src/limiter.rs (hunk 4 above)
Comment (alexcrichton) [reply crates/wasmtime/src/limiter.rs:15]: Ah ok nah I think for now this is fine. It runs a possible risk of making it difficult to optimize `memory.grow` or `table.grow` in the future but I don't think that's going to be the bottleneck of anything for a long time.

File: crates/wasmtime/src/store.rs [hunk 9]
```diff
@@ -162,10 +165,32 @@ impl Store {
...
+    /// # use wasmtime::{Engine, Store, StoreLimitsBuilder};
+    /// let engine = Engine::default();
+    /// let store = Store::new_with_limits(&engine, StoreLimitsBuilder::new().instances(10).build());
+    /// ```
+    pub fn new_with_limits(engine: &Engine, limiter: impl ResourceLimiter + 'static) -> Self {
+        let store = Store::new(engine);
+
+        {
+            let mut l = store.inner.limiter.borrow_mut();
+            *l = Some(Arc::new(ResourceLimiterProxy::new(&store, limiter)));
```
Comment (alexcrichton) [inline crates/wasmtime/src/store.rs:188]: `Arc` is used here but I think `Rc` may suffice? (I didn't see any location that seemed to require thread-safety here)

File: crates/wasmtime/src/limits.rs [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    /// This function should return `true` to indicate that the growing operation is permitted or
+    /// `false` if not permitted.
+    ///
+    /// Note that this function will be called even when the desired count exceeds the given maximum.
+    ///
+    /// Returning `true` when a maximum has been exceeded will have no effect as the linear memory
+    /// will not be grown.
+    fn memory_growing(
+        &self,
+        store: &Store,
```
Comment (alexcrichton) [inline crates/wasmtime/src/limits.rs:30]: One alternative design perhaps is that the `Store` isn't passed as an argument here and users manually `RefCell` and `Weak` themselves with the `Store`. Do you think, though, that this will be commonly desired such that we may as well go ahead and expose it?

File: tests/all/limits.rs [hunk 11]
```diff
@@ -0,0 +1,286 @@
+use anyhow::Result;
+use std::cell::{Cell, RefCell};
+use std::rc::Rc;
+use wasmtime::*;
+
+#[test]
+fn test_limits() -> Result<()> {
```
Comment (alexcrichton) [inline tests/all/limits.rs:7]: Could the tests here be expanded with tests where the initial memory/table size exceed the maximum? Additionally API requests like `Memory::new` and `Table::new` should fail as well if they exceed the maximum? (ideally also tests where API-created objects will also fail to `grow` past the maximum)

File: crates/wasmtime/src/store.rs (hunk 7 above)
Comment (alexcrichton) [reply crates/wasmtime/src/store.rs:104]: Hm ok, I had a question above about the `Store` argument and whether we can possibly remove that, but also here do you think it would make sense to remove the `Option` to effectively always have a limiting strategy?

File: crates/wasmtime/src/limits.rs (hunk 10 above)
Comment (peterhuene) [reply crates/wasmtime/src/limits.rs:30]: Okay, getting back to this PR: I guess we don't have to expose the `Store` and let them capture whatever context they want since it's a custom implementation anyway.

I'll remove the `Store` parameter and that'll simplify things a bit.

File: crates/wasmtime/src/store.rs (hunk 7 above)
Comment (peterhuene) [reply crates/wasmtime/src/store.rs:104]: The limiter needs to be an `Option` in the runtime for both table and memory so that the pooling allocator is able to "take" the memory/table when deallocating the instance (swaps in `None`).

I don't think having it also `Option` in `Store` makes it that much worse and it does have the benefit of not boxing a new trait object every time a `Store` is created just to return `true` on memory/table grow operations.

If you're okay with it, I'd like to leave this one as-is.

**Comment by peterhuene** [comment]
Only the last commit is new; the rest were from your last review prior to my rebasing this branch on main.

File: crates/runtime/src/instance/allocator/pooling.rs [hunk 12]
```diff
@@ -390,28 +426,15 @@ impl InstancePool {
...
-            )?;
-            Self::set_instance_tables(instance, self.tables.get(index), self.tables.max_elements)?;
-
-            initialize_vmcontext(instance, req);
-
-            Ok(InstanceHandle::new(instance as _))
+            self.setup_instance(index, req).or_else(|e| {
+                // Deallocate the allocated instance on error
+                let instance = self.instance(index);
+                self.deallocate(&InstanceHandle {
```
Comment (alexcrichton) [inline crates/runtime/src/instance/allocator/pooling.rs:433]: To confirm, are there tests for this? It seems a bit fishy that we're  deallocating a half-initialized instance without many modifications below. I'd expect, for example `decommit_memory_pages` to fail one one way or another if called on an othewise unmapped region?

File: crates/wasmtime/src/store.rs [hunk 13]
```diff
@@ -99,6 +101,7 @@ pub(crate) struct StoreInner {
     current_poll_cx: Cell<*mut Context<'static>>,
     out_of_gas_behavior: Cell<OutOfGas>,
     context_values: RefCell<HashMap<TypeId, Box<dyn Any>>>,
+    limiter: Option<Rc<dyn wasmtime_runtime::ResourceLimiter>>,
```
Comment (alexcrichton) [inline crates/wasmtime/src/store.rs:104]: I suppose one other possible option is to store `*mut dyn ResourceLimiter` in the `*mut VMContext` so we can use `Box` here instead of `Rc`, but let's only do that if the need actually arises.

File: crates/wasmtime/src/store.rs [hunk 14]
```diff
@@ -131,7 +134,25 @@ impl Hash for HostInfoKey {
...
+    /// Creates a new store to be associated with the given [`Engine`] and using the supplied
+    /// resource limiter.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use wasmtime::{Engine, Store, StoreLimitsBuilder};
+    /// let engine = Engine::default();
+    /// let store = Store::new_with_limits(&engine, StoreLimitsBuilder::new().instances(10).build());
+    /// ```
```
Comment (alexcrichton) [inline crates/wasmtime/src/store.rs:150]: Could the documentation here link to the `ResourceLimiter` trait which discussed which limits can be configured?

File: crates/wasmtime/src/store.rs [hunk 15]
```diff
@@ -131,7 +134,25 @@ impl Hash for HostInfoKey {
 
 impl Store {
     /// Creates a new store to be associated with the given [`Engine`].
```
Comment (alexcrichton) [inline crates/wasmtime/src/store.rs:136]: Could the documentation here be expanded to indicate that the created `Store` has no limits on usage of memory and such in wasm modules?

File: crates/wasmtime/src/store.rs (hunk 14 above)
Comment (alexcrichton) [reply crates/wasmtime/src/store.rs:150]: Oh also, could this link to `StoreLimits` and `StoreLimitsBuilder` for a built-in way to limit resources?

File: crates/wasmtime/src/limits.rs [hunk 16]
```diff
@@ -0,0 +1,207 @@
+pub(crate) const DEFAULT_INSTANCE_LIMIT: usize = 10000;
+pub(crate) const DEFAULT_TABLE_LIMIT: usize = 10000;
+pub(crate) const DEFAULT_MEMORY_LIMIT: usize = 10000;
+
+/// Used by hosts to limit resource consumption of instances at runtime.
+///
+/// A [`Store`](crate::Store) can be created with a resource limiter so that hosts can take into account
```
Comment (alexcrichton) [inline crates/wasmtime/src/limits.rs:7]: Could this link to the `new_with_limits` API?

**Review by alexcrichton — APPROVED** [review]
Looks great to me! I've got one clarification on the poolilng allocator and deinitializing halfway through, but otherwise just some documentation nits.

File: crates/runtime/src/instance/allocator/pooling.rs (hunk 12 above)
Comment (peterhuene) [reply crates/runtime/src/instance/allocator/pooling.rs:433]: The `test_pooling_allocator_initial_limits_exceeded` test does cover this, albeit with empty memory/table lists on `deallocate`, as that test is what caught the problem: with an instance limit of 1, it failed to allocate any more instances because it wasn't being added back to the free list.  I will add another memory to the module definition such that the list of initialized memories isn't empty, just in case.

While the instance isn't completely initialized, it should be safe to call `deallocate` on it as we iterate back through only initialized `Memory` and `Table` to call `decommit_memory_pages` on pages that were either made accessible (for `Memory`) or explicitly committed (for `Table`) prior to inserting into those lists.

File: crates/wasmtime/src/store.rs (hunk 13 above)
Comment (peterhuene) [reply crates/wasmtime/src/store.rs:93]: 👍 I'm going to leave this as-is and we can revisit in the the future if needed.

**Review by alexcrichton — APPROVED** [review]
